### PR TITLE
Support CEA-608 closed-captions in H262/MPEG-TS.

### PR DESCRIPTION
**Pull request by goffioul**
Issue: https://github.com/google/ExoPlayer/issues/2565

### CODE REVIEW

**Comment by AquilesCanta** [comment]
Looks nice! Thanks. Will get to this soon. Possibly next week or the one after that.

**Comment by SVPA-MaxWu** [comment]
any plan to support 708 as well?

**Comment by goffioul** [comment]
AFAIK 708 is supported, although I never tested it myself. However it's not used by default and/or it depends on whether your stream has the proper data in the stream descriptor (in case of MPEG/TS). To have 708 decoder enabled, you have to make sure that the SeiReader object is created with the appropriate mimetype: if your stream does not have appropriate descriptors and you don't provide the list of mimetypes explicitly when creating SeiReader, then it'll default to 608 only.

In all cases, this is independent from how the data is extracted from the stream, which was the point of this pull request.  608 and 708 embeds data similarly in MPEG2 and H264 streams. It's how the data is interpreted afterwards that makes a difference.

**Comment by ojw28** [comment]
@AquilesCanta - Planning to take a look at this?

File: library/core/src/main/java/com/google/android/exoplayer2/text/cea/CeaUtil.java [hunk 1]
```diff
@@ -41,7 +65,7 @@
    * @param seiBuffer The unescaped SEI NAL unit data, excluding the NAL unit start code and type.
    * @param outputs The outputs to which any samples should be written.
    */
-  public static void consume(long presentationTimeUs, ParsableByteArray seiBuffer,
+  private static void consumeSei(long presentationTimeUs, ParsableByteArray seiBuffer,
```
Comment (AquilesCanta) [inline library/core/src/main/java/com/google/android/exoplayer2/text/cea/CeaUtil.java:68]: Could we leave the javadocs for consume?

File: library/core/src/main/java/com/google/android/exoplayer2/text/cea/CeaUtil.java [hunk 2]
```diff
@@ -33,6 +37,26 @@
...
   private static final int USER_DATA_TYPE_CODE = 0x3;
 
+  public static void consume(long presentationTimeUs, ParsableByteArray buffer,
+      TrackOutput[] outputs, int mode) {
+    switch (mode) {
+      case MODE_H262:
+        consumeUserData(presentationTimeUs, buffer, outputs);
+        break;
+      case MODE_H264:
+        consumeSei(presentationTimeUs, buffer, outputs);
```
Comment (AquilesCanta) [inline library/core/src/main/java/com/google/android/exoplayer2/text/cea/CeaUtil.java:47]: Considering the modes are actually only 2 and they are always known at compile time, why not just have a consumeUserData and consumeSei made public? If so, public methods should be documented.

**Comment by okulagin** [comment]
@goffioul  any thought when comments left by @AquilesCanta will be addressed?

**Comment by goffioul** [comment]
@okulagin I don't really have the time at the moment. It should be trivial to address @AquilesCanta 's comments, though. Feel free to do it, if you wish.

**Comment by ojw28** [comment]
It looks likely support for this will be added by https://github.com/google/ExoPlayer/pull/4308.

**Pull request closed without merging**